contracts/libraries/WeightedOracleLibrary.sol

```diff
@@ -0,0 +1,81 @@
+
+        uint32[] memory secondsAgos = new uint32[](2);
+        secondsAgos[0] = period;
+        secondsAgos[1] = 0;
+
+        for (uint256 i; i < pools.length; i++) {
+            (int56[] memory tickCumulatives, uint160[] memory secondsPerLiquidityCumulativeX128s) = IUniswapV3Pool(pools[i]).observe(secondsAgos);
+            int56 tickCumulativesDelta = tickCumulatives[1] - tickCumulatives[0];
+            uint160 secondsPerLiquidityCumulativesDelta = secondsPerLiquidityCumulativeX128s[1] - secondsPerLiquidityCumulativeX128s[0];
+
+            observations[i].arithmeticMeanTick = int24(tickCumulativesDelta / period);
+            // Always round to negative infinity
+            if (tickCumulativesDelta < 0 && (tickCumulativesDelta % period != 0)) observations[i].arithmeticMeanTick--;
+
+            // We are shifting the liquidity delta to ensure that the result doesn't overflow uint128
+            observations[i].harmonicMeanLiquidity = uint128(periodX160 / (uint192(secondsPerLiquidityCumulativesDelta) << 32));
+        }
+    }
+
+    /// @notice Given some time-weighted observations, calculates the arithmetic mean tick, weighted by liquidity
+    /// @param observations A list of time-weighted observations
+    /// @return arithmeticMeanWeightedTick The arithmetic mean tick, weighted by the observations' time-weighted harmonic average liquidity
+    function getArithmeticMeanTickWeightedByLiquidity(TimeWeightedObservation[] memory observations) internal pure returns (int24 arithmeticMeanWeightedTick) {
+
+        // Accumulates the sum of all observations' products between each their own average tick and harmonic average liquidity
+        // Each product can be stored in a int160, so it would take approximatelly 2**96 observations to overflow this accumulator
+        int256 numerator;
+
+        // Accumulates the sum of the harmonic average liquidities from the given observations
+        // Each average liquidity can be stored in a uint128, so it will take approximatelly 2**128 observations to overflow this accumulator
+        uint256 denominator;
+
+        for (uint256 i; i < observations.length; i++) {
+            numerator += int256(observations[i].harmonicMeanLiquidity) * observations[i].arithmeticMeanTick;
+            denominator += observations[i].harmonicMeanLiquidity;
+        }
+
+        arithmeticMeanWeightedTick = int24(numerator / int256(denominator));
+
+        // Always round to negative infinity
+        if (numerator < 0 && (numerator % int256(denominator) != 0)) arithmeticMeanWeightedTick--;
+    }
+
+    /// @notice Given a Uniswap V3 pool, returns the timestamp of the oldest observation available
+    /// @param pool Address of Uniswap V3 pool that we want to analyze
+    /// @return oldestBlockTimestamp The timestamp of the oldest observation available on the pool
+    function getOldestObservationTimestampForPool(address pool) internal view returns (uint32 oldestBlockTimestamp) {
```

Comment:
it seems like you wouldn't ever want to sacrifice oracle security just because the data isn't available, but rather you'd just increase observation cardinality and then wait the required period for enough observations to be collected, so i would prefer not to expose this

Reply:
Ok, will revert then